### PR TITLE
MLP attention

### PR DESCRIPTION
**Pull request by pmichel31415**
Summary:
This implements the MLP style attention from the original paper.

This also reformats the `dot_attention` a little bit and adds unit tests for masking, dot and MLP attention

Differential Revision: D8811304

### CODE REVIEW

File: pytorch_translate/attention/dot_attention.py [hunk 1]
```diff
@@ -27,36 +27,22 @@ def __init__(self, decoder_hidden_state_dim, context_dim, **kwargs):
...
+        source_hids = source_hids.transpose(0, 1)
         # decoder_state: bsz x context_dim
         if self.input_proj is not None:
             decoder_state = self.input_proj(decoder_state)
-        # compute attention
-        attn_scores = (source_hids * decoder_state.unsqueeze(0)).sum(dim=2).t()
-
-        if self.src_length_masking:
-            max_src_len = source_hids.size()[0]
-            assert max_src_len == src_lengths.data.max()
```
Comment (jmp84) [inline pytorch_translate/attention/dot_attention.py]: This assert got removed in the refactoring, is this intended?

File: pytorch_translate/attention/dot_attention.py [hunk 2]
```diff
@@ -27,36 +27,22 @@ def __init__(self, decoder_hidden_state_dim, context_dim, **kwargs):
...
-                src_lengths,
-            )
-            masked_attn_scores = attn_scores.masked_fill(src_mask == 0, -np.inf)
-            # Since input of varying lengths, need to make sure the attn_scores
-            # for each sentence sum up to one
-            attn_scores = F.softmax(masked_attn_scores, dim=-1)  # bsz x srclen
-            score_denom = torch.sum(attn_scores, dim=1).unsqueeze(dim=1).expand(
-                batch_size, max_src_len
-            )
-            normalized_masked_attn_scores = torch.div(attn_scores, score_denom).t()
```
Comment (jmp84) [inline pytorch_translate/attention/dot_attention.py]: This also got removed?

File: pytorch_translate/attention/mlp_attention.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    BaseAttention,
+    register_attention,
+    attention_utils,
+)
+from pytorch_translate.common_layers import Linear
+
+
+@register_attention('mlp')
+class MLPAttention(BaseAttention):
+    """Standard MLP attention"""
```
Comment (jmp84) [inline pytorch_translate/attention/mlp_attention.py]: Could you add a link to the paper instead of "standard"?

File: pytorch_translate/attention/mlp_attention.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        )
+        self.decoder_proj = Linear(
+            decoder_hidden_state_dim, self.attention_dim, bias=False
+        )
+        self.to_scores = Linear(
+            self.attention_dim, 1, bias=False
+        )
+        self.src_length_masking = kwargs.get("src_length_masking", True)
+
+    def forward(self, decoder_state, source_hids, src_lengths):
```
Comment (jmp84) [inline pytorch_translate/attention/mlp_attention.py]: Could you put docstring here with dimensions?

File: pytorch_translate/attention/mlp_attention.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+        src_len, bsz, _ = source_hids.size()
+        # source_hids: src_len x bsz x context_dim
```
Comment (jmp84) [inline pytorch_translate/attention/mlp_attention.py]: move this to docstring and document dimensions for flat_source_hids, etc.

File: pytorch_translate/attention/dot_attention.py (hunk 2 above)
Comment (pmichel31415) [reply pytorch_translate/attention/dot_attention.py]: This was redundant. I added unit tests showing that the attention already normalizes to 1

File: pytorch_translate/attention/dot_attention.py (hunk 1 above)
Comment (pmichel31415) [reply pytorch_translate/attention/dot_attention.py]: I'm a bit ambivalent about this one. It seems like this should already be taken care of upstream (eg. in the data loader).
